### PR TITLE
Parser: Fix handling of syntax errors such as `x =- 2`

### LINKED ISSUE #7335: Engine fatal crash when adding a simple node
https://github.com/enso-org/enso/assets/4699705/8160c1eb-bbee-4c0c-955e-3cffd874517c

Trying to add a simple `_ - 2` node to take 2 off a column.

Causes the backend to crash with an assertion failed `(left == right)`.

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Most sequences of operator characters are lexed as a single operator; however, there is a [special case when the last character is `-`](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#operator-identifiers) in order to accept expressions such as `x+-1` and `x=-1`, which use a binary operator followed by a unary operator, in an unspaced subexpression.

Due to this lexical special case, special handling is needed in the parser to reject some related invalid syntaxes, such as `x +- 1` and `x =- 1`. This special handling was absent, resulting in reaching a debug assertion when such syntax errors were encountered (#7335). This PR corrects the assumption in parts of the parser that multiple consecutive operators cannot occur in unspaced subexpressions. See the newly-added test cases for examples of the rejected expressions.

### Important Notes

- Note to QA tester: I have tested that, for every `.enso` file in the repo, this PR does not change the parser's AST. If there are any regressions, they would relate to inputs that are poor style or syntax errors, both of which are not as thoroughly covered by .enso tests and library files.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -13,7 +13,6 @@ Cargo.lock
 Cargo.toml
 /lib/rust/ @MichaelMauderer @mwu-tow @farmaazon @kazcw @vitvakatu @Frizi
 /lib/rust/ensogl/ @MichaelMauderer @farmaazon @kazcw @vitvakatu @Frizi
-/lib/rust/parser/ @kazcw @jaroslavtulach
```
Comment (JaroslavTulach) [inline .github/CODEOWNERS:16]: That's removing me from the list of reviewers for changes in the parser. Shall I continue the review?

File: lib/rust/parser/debug/tests/parse.rs [hunk 2]
```diff
@@ -698,21 +700,34 @@ fn unevaluated_argument() {
...
+        ("x = y+-", block![
+            (Assignment (Ident x) "=" (OprSectionBoundary 1
+                (OprApp (Ident y) (Ok "+") (UnaryOprApp "-" ()))))]),
+        // Assign a negative number to x.
+        ("x=-1", block![(Assignment (Ident x) "=" (UnaryOprApp "-" (Number () "1" ())))]),
+        // Assign a negated value to x.
+        ("x=-y", block![(Assignment (Ident x) "=" (UnaryOprApp "-" (Ident y)))]),
     ];
-    test(&code.join("\n"), expected);
+    cases.into_iter().for_each(|(code, expected)| test(code, expected));
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/tests/parse.rs:730]: What is the benefit of using `for_each` over separating each of the tests into its own method annotated with `#[test]`? On the cons side I can envision that the `for_each` version is harder to debug.

**Review by JaroslavTulach — APPROVED** [review]

File: lib/rust/parser/debug/tests/parse.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/debug/tests/parse.rs:730]: It is easier to read. The tests are organized into Section -> Subsection -> Function -> Cases; the Function is an important unit because it tests one aspect of the functionality, and is the lowest level that can be usefully named. It's true that this is less convenient for debugging when a test fails. Sometimes I write a function or macro that supports running a batch of cases unconditionally and reporting which failed, while allowing them to be defined together without intervening per-testcase boilerplate (I have used this approach in the new fuzzy matcher). I have not bothered to do so for the parser test cases; they have failed rarely enough that my main priority for the test case API has been for the tests to serve as readable documentation of the edge cases.
